proposals/2023-08-21-utf8.md

```diff
@@ -0,0 +1,245 @@
+
+Furthermore, PromQL users will already know that it's possible to put the metric name inside the braces, for the purpose of using an operator to enable richer metric selection, like: `{__name__=~"my\.metric.*"}`. The chosen syntax serves as a shortening of this operation, and is forward-compatible with future metric matching prefix operators if we decide to implement them.
+
+### Syntax Changes
+
+We propose the following change in syntax to PromQL and the exposition format:
+
+* Label names may be quoted, either double quotes for the exposition format, or singles/doubles/backticks for PromQL.
+* A quoted string inside the label set (`{}`) without a label matching operator is default-assigned to the label name `__name__` (the metric name).
+* Two strings without a label matching operator is a syntax error.
+* Style prefers, but does not enforce, that the string without a label matching operator is the first term in the label set.
+* Strings in quotes will be unquoted using the mechanism for each language: strconv.Unquote for PromQL, and the OpenMetrics escaping mechanism for the exposition format.
+* In the exposition format, the TYPE, HELP, and UNIT lines can have double quotes around the metric name and similarly use Go quoting.
+
+### Syntax Examples:
+
+* Basic quoted metric: `{"my.dotted.metric", region="east"}`
+
+* Quoted label name: `{"my.dotted.metric", "http.status"=~"5.."}`
+
+* Classic metrics do not need to be treated differently:
+`sum(rate(regular_metric{region="east"}[5m]))`
+
+* If the user decides to put a classic metric inside the selector, it still needs to be quoted:
+`sum(rate({"regular_metric", region="east"}[5m]))`
+
+* Escape syntax if the metric has a quote: `sum(rate({"my \"quoted\" metric", region="east"}[5m]))` or use single quotes in PromQL (not available in the exposition format): `sum(rate({'my "quoted" metric', region="east"}[5m]))`
+
+* Recording Rule Syntax: `sum(rate({"my.dotted.metric:errors:rate5m", region="east"}[5m]))`. Recording rule syntax is a naming convention, not a programmatically-meaningful operator syntax.
+
+* Unicode escaping: `{"\u1F4A9"}`
+
+* Compatible with possible future Native Histogram syntax (not being proposed here):
+  `{"my.metric.with.periods", region="east"}.sum`
+
+* Compatible with possible future metric name operator syntax via prefixes (not being proposed here):
+  `{~"a\.regex.*selector.*", region="east"}` or `{!"not.this.name", region="east"}`
+
+### Unusual / Complex Characters
+
+Depending on the parsing implementation, there may be UTF-8 characters that, when encountered as native values in quotes, may cause problems with the lexers and parsers. These characters could include whitespace, especially newlines, or complex combiner characters. If such characters exist, we will require that they be escaped with `\u` rather than allow them to exist literally. As development of this proposal progresses, we will discover which characters will fall into this category.
```

Comment:
There is probably a difference between the exposition format (which is line based, so newlines have a special meaning, and (most likely) must not show up literally anywhere in names) and PromQL (where I could imagine newlines to just work). Not sure if it worth mentioning this here.